### PR TITLE
Add map and try_map methods to BoundedBTreeMap

### PR DESCRIPTION
**Pull request by benluelo**
See https://github.com/paritytech/substrate/issues/11866 for relevant discussion.

I debated on whether to have the signature of the functions passed to `map`/`try_map` pass `K` through, but I realized that this could cause the number of items in the map to decrease due to collisions (i.e. `b.map(|(_, v)| (1, v))`). While this wouldn't violate any of the guarantees of the datatype, it is odd and most likely unexpected behavior; use `.retain().map()` if that behavior is desired.

In implementing this, I was wondering if `BoundedVec` would benefit from this functionality as well. Would a trait make sense, or should it just be a method on the struct directly? I'm not sure if this operation would make sense on a `BoundedBTreeSet` for the same reason I kept the key as a reference in `BoundedBTreeMap::{map/try_map}`.

### CODE REVIEW

**Review by ggwpez — APPROVED** [review]
>Would a trait make sense, or should it just be a method on the struct directly?   
What do you think @KiChjang ? If we want to implement it for other `Bounded*` containers as well, probably a good idea, or?

File: primitives/runtime/src/bounded/bounded_btree_map.rs [hunk 1]
```diff
@@ -161,6 +161,39 @@ where
...
+	where
+		F: FnMut((&K, V)) -> T,
+	{
+		self.0
+			.into_iter()
+			.map(|(k, v)| {
+				let t = f((&k, v));
+				(k, t)
+			})
+			.try_collect()
```
Comment (KiChjang) [inline primitives/runtime/src/bounded/bounded_btree_map.rs:176]: Why `try_collect` and `unwrap()` when you can just `collect`?

**Comment by KiChjang** [comment]
> > Would a trait make sense, or should it just be a method on the struct directly?
> 
> What do you think @KiChjang ? If we want to implement it for other `Bounded*` containers as well, probably a good idea, or?

I don't foresee anything _wrong_ with implementing it on bounded types. Perhaps your hesitation comes from the fact that we might change the size of the elements, and hence the bounded size of the overall map? This wouldn't be an issue in my opinion, as the function returns a brand new `BoundedBTreeMap`, and so can be treated as a completely separate entity as the map that was operated over.

File: primitives/runtime/src/bounded/bounded_btree_map.rs (hunk 1 above)
Comment (benluelo) [reply primitives/runtime/src/bounded/bounded_btree_map.rs]: Good point, fixed in https://github.com/paritytech/substrate/pull/11869/commits/2a993a48d47148d3f7db2ce577c7d293f00fd76b

**Comment by ggwpez** [comment]
> I don't foresee anything _wrong_ with implementing it on bounded types. Perhaps your hesitation comes from the fact that we might change the size of the elements, and hence the bounded size of the overall map? This wouldn't be an issue in my opinion, as the function returns a brand new `BoundedBTreeMap`, and so can be treated as a completely separate entity as the map that was operated over.

Just wanted to hear your opinion, thanks :smile:   
I guess we can go ahead with this?

**Comment by KiChjang** [comment]
Wait, I just realized... the `map`/`try_map` method are implemented directly as methods on the `BoundedBTreeMap` type, and I don't think this is required. We could've just implemented a `iter_mut()` method and gotten all of the `map` methods for free.

So unfortunately no, I don't think this is actually the right approach to the problem.

**Comment by benluelo** [comment]
Added iter_mut, let me know what you think

**Comment by ggwpez** [comment]
So `m.map(|(k, v)| 123)` now becomes `m.iter_mut().for_each(|(k, v)| *k = 123)` and for `try_` respectively?  
Fair enough, then we dont need the dedicated `map` and `try_map` functions.

File: primitives/runtime/src/bounded/bounded_btree_map.rs [hunk 2]
```diff
@@ -161,6 +161,41 @@ where
...
+	{
+		Ok(BoundedBTreeMap::<K, T, S>::unchecked_from(
+			self.0
+				.into_iter()
+				.map(|(k, v)| (f((&k, v)).map(|t| (k, t))))
+				.collect::<Result<BTreeMap<_, _>, _>>()?,
+		))
+	}
+
+	pub fn iter_mut(&mut self) -> sp_std::collections::btree_map::IterMut<K, V> {
```
[the remark this answers is not in the document]
Comment (benluelo) [reply primitives/runtime/src/bounded/bounded_btree_map.rs:168]: Basically just copied the docs from BTreeMap::iter_mut, let me know if I should add anything else :+1:

**Review by shawntabrizi — APPROVED** [review]

**Review by KiChjang — APPROVED** [review]

**Comment by KiChjang** [comment]
bot merge

**Comment by benluelo** [comment]
One issue with `iter_mut` is that it doesn't allow you to transform the item's *type* in the collection - you can alter the value, but if you want to go from a `Map<A, B>` to a `Map<A, C>` you still need to into_iter/map/try_collect/unwrap (which is not ideal if they're bounded by the same `S`). I think there's usecases for both, personally.

**Pull request merged**